btcutil/psbt/taproot.go

```diff
@@ -0,0 +1,219 @@
+package psbt
+
+import (
+	"bytes"
+
+	"github.com/btcsuite/btcd/btcec/v2/schnorr"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+)
+
+const (
+	// schnorrSigMinLength is the minimum length of a Schnorr signature
+	// which is 64 bytes.
+	schnorrSigMinLength = schnorr.SignatureSize
+
+	// schnorrSigMaxLength is the maximum length of a Schnorr signature
+	// which is 64 bytes plus one byte for the appended sighash flag.
+	schnorrSigMaxLength = schnorrSigMinLength + 1
+)
+
+// TaprootScriptSpendSig encapsulates an individual Schnorr signature for a
+// given public key and leaf hash.
+type TaprootScriptSpendSig struct {
+	XOnlyPubKey []byte
+	LeafHash    []byte
+	Signature   []byte
+	SigHash     txscript.SigHashType
+}
+
+// checkValid checks that both the pubkey and the signature are valid.
+func (s *TaprootScriptSpendSig) checkValid() bool {
+	return validateXOnlyPubkey(s.XOnlyPubKey) &&
+		validateSchnorrSignature(s.Signature)
+}
+
+// EqualKey returns true if this script spend signature's key data is the same
+// as the given script spend signature.
+func (s *TaprootScriptSpendSig) EqualKey(other *TaprootScriptSpendSig) bool {
+	return bytes.Equal(s.XOnlyPubKey, other.XOnlyPubKey) &&
+		bytes.Equal(s.LeafHash, other.LeafHash)
+}
+
+// TaprootScriptSpendSigSorter implements sort.Interface for the
+// TaprootScriptSpendSig struct.
+type TaprootScriptSpendSigSorter []*TaprootScriptSpendSig
+
+func (s TaprootScriptSpendSigSorter) Len() int { return len(s) }
+
+func (s TaprootScriptSpendSigSorter) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+func (s TaprootScriptSpendSigSorter) Less(i, j int) bool {
+	return bytes.Compare(s[i].XOnlyPubKey, s[j].XOnlyPubKey) < 0 &&
+		bytes.Compare(s[i].LeafHash, s[j].LeafHash) < 0
+}
+
+// TaprootTapLeafScript represents a single taproot leaf script that is
+// identified by its control block.
+type TaprootTapLeafScript struct {
+	ControlBlock []byte
+	Script       []byte
+	LeafVersion  txscript.TapscriptLeafVersion
+}
+
+// checkValid checks that the control block is valid.
+func (s *TaprootTapLeafScript) checkValid() bool {
+	return validateControlBlock(s.ControlBlock)
+}
+
+// TaprootTapLeafScriptSorter implements sort.Interface for the
+// TaprootTapLeafScript struct.
+type TaprootTapLeafScriptSorter []*TaprootTapLeafScript
+
+func (s TaprootTapLeafScriptSorter) Len() int { return len(s) }
+
+func (s TaprootTapLeafScriptSorter) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
```

Comment:
FWIW if we want to cut down on the extra/duplicate definitions a bit, we can use the `sort.Slice` function instead: https://pkg.go.dev/sort#Slice. So the only thing we'd need specifically for each new slice type is the comparison function below.

Reply:
Yeah, makes sense. Went with an approach that adds a `SortBefore` method to the structs so the business logic can be implemented on the struct itself and not where we do the actual sort.